Clean up code and add several minor features

apcraig opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    Clean up code and add a few new features
- [X] Developer(s): 
    apcraig
- [X] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    Tested with "clean01" in Icepack, https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#e630284d5baa877f93c8ca3003bcb2fa8c27e63f
- How much do the PR code changes differ from the unmodified code? 
    - [X] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [X] Yes
    - [ ] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [ ] No 
- [X] Please provide any additional information or relevant details below:

- Update debug blocks output, closes #718 
- Add memory use batch setting for machines, closes #674 
- Migrate batch task/thread computations to a new file, setup_mach_params.csh, closes #650 
- Update subroutine diagnostic_abort which calls print_state for a failing point, added to departure points errors and a few other places, closes #622 
- Update the miniconda install information so we no longer download to ~/Downloads on MacOS, closes #547
- Check conformation with f2003 and f2008.  We conform to 2003 except use of "contiguous" in 1d evp solver.  That is f2008.  Also compiled with -Wall and removed some unused variables.  Closes #455 
- Remove trailing blank space via a automated script.
- Fixed automated qc testing on cheyenne for prod_suite testsuite.  Was not completing the qc test completely.  Needed to create a conda env to install utilities for the qc tool.  The old python env we were using was deprecated recently.  The new NCAR npl conda env didn't work because it broke the CICE compile.  pip install didn't work for compatibility reasons not well understood.  So I added a qctest.yml file to configuration/scripts/tests and installed it in my env by default.  This seems to fix the problem for me, but use of the automated qc testing defined in the prod_suite may cause problems for other users and on other machines.

apcraig commented:
@dabail10, could you have a quick look and approve.  I know there is a lot here.  Would like to merge today for weekend testing if possible.  Thanks!

phil-blain commented:
I'll also have a look.

apcraig commented:
Thanks @phil-blain.  In hindsight, I shouldn't have put in all the code cleanup (especially cleaning trailing blanks) with the other mods.  I could split up this PR if others feel it's important.

phil-blain commented on cicecore/cicedynB/dynamics/ice_transport_remap.F90:
```diff
@@ -832,15 +826,7 @@ subroutine horizontal_remap (dt,             ntrace,   &
                              mm   (:,:,0,iblk))
 
          if (l_stop) then
-            this_block = get_block(blocks_ice(iblk),iblk)
-            write (nu_diag,*) 'istep1, my_task, iblk, cat =',     &
-                               istep1, my_task, iblk, '0'
```
here also we loose this `cat = 0` bit of info, but I'm not sure if it was really useful...

apcraig replied:
Agree, I don't think this adds much so will not re-add.

phil-blain commented on cicecore/cicedynB/dynamics/ice_transport_remap.F90:
```diff
@@ -858,14 +844,7 @@ subroutine horizontal_remap (dt,             ntrace,   &
                                tm    (:,:,:,n,iblk))
 
             if (l_stop) then
-               write (nu_diag,*) 'istep1, my_task, iblk, cat =',     &
-                                  istep1, my_task, iblk, n
```
and again here we loose the category info

apcraig replied:
I am adding back the category output.

phil-blain commented on doc/source/user_guide/ug_running.rst:
```diff
@@ -535,10 +535,10 @@ On macOS:
-  # Download the Miniconda installer to ~/Downloads/miniconda.sh
-  curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh -o ~/Downloads/miniconda.sh
+  # Download the Miniconda installer to ~/miniconda.sh
+  curl -L https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh -o ~/miniconda.sh
   # Install Miniconda
-  bash ~/Downloads/miniconda.sh
+  bash ~/miniconda.sh
```
Nice to take the time to fix this. If you update your commit message to "Closes https://github.com/CICE-Consortium/CICE/issues/547" instead of "See https://github.com/CICE-Consortium/CICE/issues/547", the issue will be closed automatically when your PR is merged :) 

That could be done for the other issues you mention in previous commits, also, if you wish to close these issues.

apcraig replied:
Just to clarify, I added the "closes" to the PR documentation.  I think this is a good place for it and how I usually do it.  Interesting to know that similar syntax in the commit message could do the same thing.

phil-blain commented on cicecore/cicedynB/dynamics/ice_dyn_vp.F90:
```diff
@@ -87,7 +88,8 @@ module ice_dyn_vp
 
       character (len=char_len), public :: &
-         precond        , & ! preconditioner for fgmres: 'ident' (identity), 'diag' (diagonal), 'pgmres' (Jacobi-preconditioned GMRES)
+         precond        , & ! preconditioner for fgmres: 'ident' (identity), 'diag' (diagonal), 'pgmres'
+                            ! (Jacobi-preconditioned GMRES)
```
minor, but I'd prefer keeping the "pgmres' option on the same line as its description

apcraig replied:
This is one the few things -Wall complained about.  Not sure what the "extended" line limit is, but some of these lines were "too long".  I'm happy to undo all of these or is it better to conform to what -Wall says?

apcraig replied:
this was fixed.

phil-blain commented on cicecore/cicedynB/general/ice_init.F90:
```diff
@@ -409,7 +413,8 @@ subroutine input_data
       capping_method  = 'max'  ! method for capping of viscosities (max=Hibler 1979,sum=Kreyscher2000)
       maxits_nonlin = 4        ! max nb of iteration for nonlinear solver
-      precond = 'pgmres'       ! preconditioner for fgmres: 'ident' (identity), 'diag' (diagonal), 'pgmres' (Jacobi-preconditioned GMRES)
+      precond = 'pgmres'       ! preconditioner for fgmres: 'ident' (identity), 'diag' (diagonal), 'pgmres'
+                               ! (Jacobi-preconditioned GMRES)
```
same here

apcraig replied:
fixed

phil-blain commented on cicecore/cicedynB/dynamics/ice_dyn_vp.F90:
```diff
@@ -69,7 +69,8 @@ module ice_dyn_vp
          maxits_fgmres  , & ! max nb of iteration for fgmres
          maxits_pgmres  , & ! max nb of iteration for pgmres
-         fpfunc_andacc  , & ! fixed point function for Anderson acceleration: 1: g(x) = FMGRES(A(x),b(x)), 2: g(x) = x - A(x)x + b(x)
+         fpfunc_andacc  , & ! fixed point function for Anderson acceleration: 1:
+                            ! g(x) = FMGRES(A(x),b(x)), 2: g(x) = x - A(x)x + b(x)
```
the `1: ` should be on the second line.

apcraig replied:
fixed

phil-blain commented on cicecore/cicedynB/general/ice_init.F90:
```diff
@@ -422,7 +427,8 @@ subroutine input_data
       algo_nonlin = 'picard'        ! nonlinear algorithm: 'picard' (Picard iteration), 'anderson' (Anderson acceleration)
-      fpfunc_andacc = 1        ! fixed point function for Anderson acceleration: 1: g(x) = FMGRES(A(x),b(x)), 2: g(x) = x - A(x)x + b(x)
+      fpfunc_andacc = 1        ! fixed point function for Anderson acceleration: 1: 
+                               ! g(x) = FMGRES(A(x),b(x)), 2: g(x) = x - A(x)x + b(x)
```
the `1: ` should be on the second line  here too

apcraig replied:
fixed

phil-blain reviewed (changes requested):
Very nice clean-up in general, thanks a lot for these! A few comments on the commits themselves:
- 58e3b8f `Update/improve debug_blocks output, see #718.`
  - thanks!
- 3b0d98d `Add ICE_MEMUSE cice.settings flag for batch memory use Add set_env.memsmall, memmed, memlarge options To use, will require changes to the env machine files.  Most machines will probably not use it. See #674.`
  - thanks!
- 6abb59b `Add setup_machparams.csh to compute batch/launch machine parameters Update cice.batch.csh and cice.launch.csh to use setup_machparams.csh See #650`
  - thanks a lot for that, it's a nice improvement!
- a14eb91 `Update subroutine diagnostic_abort which calls print_state Update ice_transport_remap and ice_transport_driver to call diagnostic_abort   during some errors. See also #622`
  - see comments inline
- f6d09e4 `Update miniconda install information See #547`
  - idem
- a95c35c `Code cleanup based on compile with -Wall Code cleanup based on -std f2003 and f2008 checks Add -stand f08 to cheyenne_intel debug flags Add -std f2008 to cheyenne_gnu debug flags Code consistent with Fortran 2003 except for use of contiguous in   1d evp code.`
  - nice to see we are conformant!
- d3b2294 `Remove all trailing blank space with script`
- e630284 `Update the cheyenne env so qc testing works Add configuration/scripts/tests/qctest.yml file Update documentation`

Note: this is the output of `git log --oneline --reverse  --no-decorate upstream/main..apcraig/debug01`. As you can see Git parses your commit messages as a single line, because they are missing a blank line between the first line (subject) and the rest (body). It would be nice to follow this convention, as I'm not sure how much the squash merge message generated by GitHub will reformat these messages. It's also a good idea also to limit each line to 70 characters for a nice reading experience on the terminal. (Just my 2 cents ;))

dabail10 reviewed (approved):
You can actually review changes by commit. There is a pull-down menu on the left.

dabail10 commented on cicecore/cicedynB/general/ice_flux_bgc.F90:
```diff
@@ -138,22 +138,22 @@ subroutine alloc_flux_bgc
+         HDO_ocn     (nx_block,ny_block,max_blocks), & ! seawater concentration of HDO (kg/kg)
+         H2_16O_ocn  (nx_block,ny_block,max_blocks), & ! seawater concentration of H2_16O (kg/kg)
+         H2_18O_ocn  (nx_block,ny_block,max_blocks), & ! seawater concentration of H2_18O (kg/kg)
+         Qa_iso      (nx_block,ny_block,icepack_max_iso,max_blocks), & ! isotope specific humidity (kg/kg)
```
Did you reintroduce trailing spaces here?

apcraig replied:
I just shifted over the () to align better.  The blank was added to the middle of the line.

dabail10 commented:
I had a quick look through everything and didn't see anything beyond @phil-blain 's comprehensive review.

apcraig commented on cicecore/cicedynB/dynamics/ice_transport_driver.F90:
```diff
@@ -688,9 +688,7 @@ subroutine transport_remap (dt)
                                         istop,              jstop)
 
                if (ckflag) then
-                  write (nu_diag,*) 'istep1, my_task, iblk, cat =',     &
-                                     istep1, my_task, iblk, n
-                  call abort_ice(subname//'ERROR: monotonicity error')
```
I'll look into the diagnostic changes and fix them.  Thanks for catching that.

apcraig commented on cicecore/cicedynB/dynamics/ice_transport_driver.F90:
```diff
@@ -688,9 +688,7 @@ subroutine transport_remap (dt)
-                  write (nu_diag,*) 'istep1, my_task, iblk, cat =',     &
-                                     istep1, my_task, iblk, n
-                  call abort_ice(subname//'ERROR: monotonicity error')
+                  call diagnostic_abort(istop,jstop,iblk,' monotonicity error')
```
I have added back the category output.

apcraig commented on cicecore/cicedynB/analysis/ice_diagnostics.F90:
```diff
@@ -2120,18 +2115,23 @@ subroutine diagnostic_abort(istop, jstop, iblk, istep1, stop_label)
+         write (nu_diag,*) subname,' Global block:', this_block%block_id
+         write (nu_diag,*) subname,' Global i and j:', &
+                             this_block%i_glob(istop), &
+                             this_block%j_glob(jstop) 
```
Good catch, I fixed this.

apcraig commented:
I fixed the output noted by @phil-blain and ran quick suites on cheyenne to confirm the changes compile and run fine.  I also updated Icepack to the latest version which also has similar cleanup.  I was testing this branch with that version already, so I think it's all fine.

phil-blain commented on cicecore/cicedynB/analysis/ice_diagnostics.F90:
```diff
@@ -2118,20 +2109,17 @@ subroutine diagnostic_abort(istop, jstop, iblk, istep1, stop_label)
+      else
+         write (nu_diag,*) subname,' istep1, my_task, iblk =', &
+                            istep1, my_task, iblk
+         write (nu_diag,*) subname,' Global block:', this_block%block_id
```
In the old code the global block ID was always printed, here it is only printed if `istop` and `jstop` are not both greater than zero, i.e. if we go through the `else` branch.

I think "Global block" should also be printed by `print_state` to keep the same amount of info. Sorry I missed that on my first review.

apcraig replied:
I'm adding it now.  Good idea.

phil-blain commented on cicecore/cicedynB/dynamics/ice_transport_driver.F90:
```diff
@@ -689,6 +689,8 @@ subroutine transport_remap (dt)
 
                if (ckflag) then
+                  write (nu_diag,*) 'istep1, my_task, iblk, cat =',     &
+                                     istep1, my_task, iblk, n
```
OK, here we will end up with duplicated printing of `istep1, my_task, iblk`, but we do not loose the category info. That's OK.
Out of curiosity, why not add the category as an argument to `diagnostic_abort` and `print_state` ? I'm guessing the answer is "to not have to update all callers of these subroutine", which I think does make sense.

apcraig commented:
I added global block output to print state and ran a handful of tests on cheyenne to make sure it built and ran.
